gems/smithy-xml/lib/smithy-xml/parser/frame.rb

```diff
@@ -0,0 +1,341 @@
+        end
+
+        def consume_child_frame(child); end
+
+        # @api private
+        def path
+          if parent.is_a?(Stack)
+            [@path]
+          else
+            parent.path + [@path]
+          end
+        end
+
+        # @api private
+        def yield_unhandled_value(path, value)
+          parent.yield_unhandled_value(path, value)
+        end
+      end
+
+      # @api private
+      class BigDecimalFrame < Frame
+        def result
+          @text.empty? ? nil : BigDecimal(@text.join)
+        end
+      end
+
+      # @api private
+      class BlobFrame < Frame
+        def result
+          @text.empty? ? '' : Base64.decode64(@text.join)
+        end
+      end
+
+      # @api private
+      class BooleanFrame < Frame
+        def result
+          @text.empty? ? nil : (@text.join == 'true')
+        end
+      end
+
+      # @api private
+      class IntegerFrame < Frame
+        def result
+          @text.empty? ? nil : @text.join.to_i
+        end
+      end
+
+      # @api private
+      class FlatListFrame < Frame
+        def initialize(xml_name, *args)
+          super
+          @member = Frame.new(xml_name, self, @ref.shape.member)
+        end
+
+        def result
+          @member.result
+        end
+
+        def append_text(value)
+          @member.append_text(value)
+        end
+
+        def child_frame(xml_name)
+          @member.child_frame(xml_name)
+        end
+
+        def consume_child_frame(_child)
+          @result = @member.result
+        end
+      end
+
+      # @api private
+      class FloatFrame < Frame
+        def result
+          @text.empty? ? nil : deserialize_number(@text.join)
+        end
+
+        # @param [String] str
+        # @return [Number] The input as a number
+        def deserialize_number(str)
+          case str
+          when 'Infinity' then ::Float::INFINITY
+          when '-Infinity' then -::Float::INFINITY
+          when 'NaN' then ::Float::NAN
+          when nil then nil
+          else str.to_f
+          end
+        end
+      end
+
+      # @api private
+      class ListFrame < Frame
+        def initialize(*args)
+          super
+          @result = []
+          @member_xml_name = @ref.shape.member.traits['smithy.api#xmlName'] || 'member'
+        end
+
+        def child_frame(xml_name)
+          unless xml_name == @member_xml_name
+            raise NotImplementedError, "Expected XML name '#{@member_xml_name}' for ListFrame, got '#{xml_name}'"
+          end
+
+          Frame.new(xml_name, self, @ref.shape.member)
+        end
+
+        def consume_child_frame(child)
+          @result << child.result unless child.is_a?(NullFrame)
+        end
+      end
+
+      # @api private
+      class MapEntryFrame < Frame
+        def initialize(xml_name, *args)
+          super
+          @key_name = @ref.shape.key.traits['smithy.api#xmlName'] || 'key'
+          @key = Frame.new(xml_name, self, @ref.shape.key)
```

Comment:
Supposed to be StringFrame here per documenation?

Reply:
Yes. It ends up being a StringFrame. Frame.new will look at the shape and make the appropriate class.